### PR TITLE
feat(website): download sequences with customized fasta headers

### LINKED ISSUE #2284: ENH: Add display name after sequence id in downloads, or allow customization of fasta id, so it can contain metadata or display name
<strike>_Blocked on LAPIS: https://github.com/GenSpectrum/LAPIS/issues/857_</strike>

I downloaded the most recent mpox sequences from loculus full-mpox to put them into Nextclade.

Within Nextclade, I noticed that all I got is the loculus accession, and I'm sorely missing metadata - i.e. what's currently in our display name.

Would be great if we could either always add the display name after a separator, e.g. `LOCACCESSION|DISPLAYNAME`

Or allow the user to configure the fasta id - I guess that's hard if we use LAPIS as LAPIS probably doesn't have a concept of multiple or configurable sequence ids - unless I'm missing something.

Brave Browser 2024-07-10 16 47 21

Alternatives for searchability: customize adding metadata to fasta sequence headers

### PR DESCRIPTION
**Pull request by fengelniederhammer**
resolves #2284

preview URL:

### Summary

This adapts @corneliusroemer's work of #3448 (the branch had too many ugly merge conflicts, so I started from scratch copying the changes from the other branch). It also adapts the Astro endpoint to stream the resulting fasta. It first loads a map of metadata of all sequences into memory (to get a map "accession version -> fasta header"). Then it fetches the sequences, but directly forwards the stream while mapping the fasta headers.

### PR Checklist

- [ ] All necessary documentation has been adapted.
- [ ] The implemented feature is covered by appropriate, automated tests.
- [ ] Any manual testing that has been done is documented (i.e. what exactly was tested?)

### CODE REVIEW

**Comment by fengelniederhammer** [comment]
I believe that it mostly works at this stage, you're welcome to have a look. A couple of things are still missing (edge cases, tests, docs). I'll check on those on Monday.

File: website/src/pages/[organism]/api/sequences.ts [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+        ...searchParams.queryFilters,
+        fields: [ACCESSION_VERSION_FIELD, ...searchParams.headerFields],
+    });
+
+    return details.map((it) => {
+        const fastaHeaderMap = new Map<string, string>();
+
+        for (const datum of it.data) {
+            const { [ACCESSION_VERSION_FIELD]: accessionVersion, ...rest } = datum;
+            const fastaHeader = Object.values(rest).join('/');
```
Comment (chaoran-chen) [inline website/src/pages/[organism]/api/sequences.ts:93]: Let's rather join with `|`, that's also what NCBI Virus uses.

```suggestion
            const fastaHeader = Object.values(rest).join('|');
```

File: kubernetes/loculus/values.yaml
Comment (chaoran-chen) [inline kubernetes/loculus/values.yaml:1]: The Helm config schema needs to be extended.

(@fhennig, now that #3770 has been fixed, shouldn't the helm-schema-lint action have failed?)

File: website/src/components/SearchPage/DownloadDialog/DownloadUrlGenerator.ts
Comment (chaoran-chen) [inline website/src/components/SearchPage/DownloadDialog/DownloadUrlGenerator.ts:1]: When I first select a compression and then switch the header style, the download leads to a 500 error.

File: website/src/components/SearchPage/DownloadDialog/DownloadForm.tsx [hunk 2]
```diff
@@ -87,19 +90,31 @@ export const DownloadForm: FC<DownloadFormProps> = ({
...
+                                  name='unalignedNucleotideSequences'
+                                  options={referenceGenomesSequenceNames.nucleotideSequences.map((segment) => ({
+                                      label: <>{segment}</>,
+                                  }))}
+                                  selected={unalignedNucleotideSequence}
+                                  onSelect={setUnalignedNucleotideSequence}
+                                  disabled={dataType !== 1}
+                              />
+                          ) : undefined}
+                          {/* TODO: probably hide this when not set in values.yml? */}
```
Comment (chaoran-chen) [inline website/src/components/SearchPage/DownloadDialog/DownloadForm.tsx:106]: Yes, agree that this should be hidden if not set.

File: website/src/pages/[organism]/api/sequences.ts [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+
+    const { fastaHeaderMap, dataVersion: detailsDataVersion } = fastaHeaderMapResult.value;
+
+    const sequencesResponse = await lapisClient.streamSequences(searchParams.segment, {
+        ...searchParams.queryFilters,
+        dataFormat: 'ndjson',
+    });
+
+    const sequencesDataVersion = sequencesResponse.headers['lapis-data-version'];
+    if (sequencesDataVersion !== detailsDataVersion) {
```
Comment (chaoran-chen) [inline website/src/pages/[organism]/api/sequences.ts:42]: It would be better to retry (once?) in this case, but as a temporary solution, we can leave it as is.

File: website/src/components/SearchPage/DownloadDialog/DownloadUrlGenerator.ts [hunk 4]
```diff
@@ -86,4 +100,21 @@ export class DownloadUrlGenerator {
...
+
+    private downloadEndpoint(dataType: DownloadDataType) {
+        const segmentPath = (segment?: string) => (segment !== undefined ? '/' + segment : '');
+
+        switch (dataType.type) {
+            case 'metadata':
+                return this.lapisUrl + '/sample/details';
+            case 'unalignedNucleotideSequences':
+                return dataType.includeRichFastaHeaders === true
+                    ? '/' + this.organism + '/api/sequences'
```
Comment (chaoran-chen) [inline website/src/components/SearchPage/DownloadDialog/DownloadUrlGenerator.ts:112]: This should be a full URL because the value is also used when a user uses the "Copy download URL" button.

File: website/src/pages/[organism]/api/sequences.ts [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+                    if (line.trim() === '') {
+                        continue;
+                    }
+
+                    try {
+                        const data = ndjsonLineSchema.parse(JSON.parse(line));
+
+                        const fastaHeader = fastaHeaderMap.get(data.accessionVersion as string);
+                        if (!fastaHeader) {
+                            const reason = `Did not find fasta header for accession version ${data.accessionVersion}`;
```
Comment (chaoran-chen) [inline website/src/pages/[organism]/api/sequences.ts:148]:
```suggestion
                            const reason = `Did not find metadata for accession version ${data.accessionVersion}`;
```

Would this be clearer? Or maybe point out that the details is missing values?

File: website/src/pages/[organism]/api/sequences.ts [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+        },
+    });
+};
+
+function getSearchParams(url: URL) {
+    const searchParams = url.searchParams;
+
+    const segment = searchParams.get('segment') ?? undefined;
+    const downloadFileBasename = searchParams.get('downloadFileBasename') ?? 'sequences';
+    const headerFields = searchParams.getAll('headerFields');
```
Comment (chaoran-chen) [inline website/src/pages/[organism]/api/sequences.ts:69]: Should we throw an error if `headerFields` is empty?

File: kubernetes/loculus/templates/_common-metadata.tpl [hunk 7]
```diff
@@ -175,6 +175,9 @@ organisms:
       {{- with ($instance.schema | include "loculus.patchMetadataSchema" | fromYaml) }}
       organismName: {{ quote .organismName }}
       loadSequencesAutomatically: {{ .loadSequencesAutomatically | default false }}
+      {{ if .richFastaHeaderFields}}
+      richFastaHeaderFields: {{ toJson .richFastaHeaderFields }}
+      {{ end }}
```
Comment (chaoran-chen) [inline kubernetes/loculus/templates/_common-metadata.tpl:180]: How should we treat an empty array? Should we treat it as equivalent to not having set it at all?

**Review by chaoran-chen — COMMENTED** [review]
Thanks very much, @fengelniederhammer, it looks already really good to me!

File: kubernetes/loculus/templates/_common-metadata.tpl (hunk 7 above)
Comment (chaoran-chen) [reply kubernetes/loculus/templates/_common-metadata.tpl:180]: Sounds good as well! Can we define that in the schema?

File: kubernetes/loculus/values.yaml
Comment (fengelniederhammer) [reply kubernetes/loculus/values.yaml:1]: Does it maybe not fail because it's in the `defaultOrganismConfig`? I checked it: `helm lint .` doesn't fail when `richFastaHeaderFields` is missing in the schema.

File: kubernetes/loculus/templates/_common-metadata.tpl (hunk 7 above)
Comment (fengelniederhammer) [reply kubernetes/loculus/templates/_common-metadata.tpl:180]: The JSON schema has an option `"minItems": 1` that let's us do that.

File: website/src/components/SearchPage/DownloadDialog/DownloadUrlGenerator.ts
Comment (fengelniederhammer) [reply website/src/components/SearchPage/DownloadDialog/DownloadUrlGenerator.ts:1]: Should be fixed now. I forgot to ignore compression in that case.

File: kubernetes/loculus/values.yaml
Comment (fhennig) [reply kubernetes/loculus/values.yaml:1]: That should also be schecked with a shema:

```
"defaultOrganismConfig": { "$ref": "#/definitions/organism" },
```

I think I missed something, that particular object doesn't have `additionalProperties: false` set. Maybe @fengelniederhammer  you could add it in? Sorry for that!

Comment (fengelniederhammer) [reply kubernetes/loculus/values.yaml:1]: I'll leave that to #3815 as it's more than a 1 minute change it seems.

File: website/src/components/SearchPage/DownloadDialog/DownloadForm.tsx [hunk 8]
```diff
@@ -87,19 +93,32 @@ export const DownloadForm: FC<DownloadFormProps> = ({
...
+                                  selected={unalignedNucleotideSequence}
+                                  onSelect={setUnalignedNucleotideSequence}
+                                  disabled={dataType !== 1}
+                              />
+                          ) : undefined}
+                          {richFastaHeaderFields && (
+                              <RadioOptionBlock
+                                  name='richFastaHeaders'
+                                  title='FASTA header style'
+                                  options={[{ label: <>Accession</> }, { label: <>Display name</> }]}
```
Comment (chaoran-chen) [inline website/src/components/SearchPage/DownloadDialog/DownloadForm.tsx:113]: We shouldn't hardcode "Display name" but due to limitation of time, I think it's fine to merge as is and address it in a follow-up issue: #3818

File: website/src/pages/[organism]/api/sequences.ts
Comment (chaoran-chen) [inline website/src/pages/[organism]/api/sequences.ts:1]: If I filter for many accession versions so that a GET request doesn't work anymore but a POST is needed, I get redirected to `https://fabian-fastaheaders.loculus.org/ebola-sudan/api/sequences` with a Page not found error.

I guess it's easiest if we disable the option if the URL is too long so that a POST request would be needed?

Comment (theosanderson) [reply website/src/pages/[organism]/api/sequences.ts:1]: (and then maybe track in an issue as this would be a nice-to-have eventually)

Comment (fengelniederhammer) [reply website/src/pages/[organism]/api/sequences.ts:1]: I disabled it.

File: website/src/pages/[organism]/api/sequences.ts [hunk 9]
```diff
@@ -0,0 +1,280 @@
...
+    });
+
+    const shouldContainAccessionVersion = searchParams.headerFields.includes(ACCESSION_VERSION_FIELD);
+
+    return details.map((it) => {
+        const fastaHeaderMap = new Map<string, string>();
+
+        for (const datum of it.data) {
+            const { [ACCESSION_VERSION_FIELD]: accessionVersion, ...rest } = datum;
+            const fastaHeader = Object.values(shouldContainAccessionVersion ? datum : rest).join('|');
```
[the remark this answers is not in the document]
Comment (chaoran-chen) [reply website/src/pages/[organism]/api/sequences.ts:188]: I don't think so. I tried `https://fabian-fastaheaders.loculus.org/cchf/api/sequences?downloadFileBasename=cchf_nuc-L_2025-03-10T1551&versionStatus=LATEST_VERSION&isRevocation=false&dataUseTerms=OPEN&headerFields=displayName&headerFields=submittedDate&headerFields=authors&segment=L` and got 

```
>Hewson, R.; Gmyl, A.; Gmyl, L.; Smirnova, S. E.; Karganova, G.; Jamil, B.; Hasan, R.; Chamberlain, J.; Clegg, C.|LOC_000M4ET.1|2025-03-10
```

I.e., the author is the first.

**Review by chaoran-chen — COMMENTED** [review]
Thank you very much! Apart from the one open comment regarding the field order, everything looks good to me! (But it would be great if a second person could take a look. @theosanderson?)

File: website/src/pages/[organism]/api/sequences.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (fengelniederhammer) [reply website/src/pages/[organism]/api/sequences.ts]: I thought LAPIS already does that. But I think we only implemented that for CSV.

I fixed it.

**Review by theosanderson — APPROVED** [review]

**Comment by theosanderson** [comment]
It looks nice to me from a quick look over. I think eventually we should change some stuff, e.g. not requiring `headerFields` to be set (defaulting to just accessionVersion or something), but we can do any of that later.

File: website/src/components/SearchPage/DownloadDialog/DownloadButton.tsx [hunk 10]
```diff
@@ -90,7 +101,7 @@ export const DownloadButton: FC<DownloadButtonProps> = ({
     return (
         <div className='flex items-center'>
             <a
-                className={`btn loculusColor ${disabled ? 'btn-disabled' : ''} text-white`}
+                className={`btn loculusColor ${disabled || handleClick === undefined ? 'btn-disabled' : ''} text-white`}
```
Comment (theosanderson) [inline website/src/components/SearchPage/DownloadDialog/DownloadButton.tsx:104]: I think we probably need a message here ideally saying "Sorry, we don't yet support displayName downloads for long queries" or something

**Review by chaoran-chen — APPROVED** [review]
Nice, thanks for fixing the orders!

File: website/src/components/SearchPage/DownloadDialog/DownloadButton.tsx (hunk 10 above)
Comment (fengelniederhammer) [reply website/src/components/SearchPage/DownloadDialog/DownloadButton.tsx:104]: Sounds good. I'll do that in a separate PR.
